### PR TITLE
[sk] Added template loading engine

### PR DESCRIPTION
**Pull request by skunichetty**
# Summary
Added template loading engine to pick the right template from file and write it to Block code upon creation.
This involves updating the  payload for adding blocks with the `config` attribute that holds template loading configuration.
```json
{
    "block": {
        "name": "name",
	"type": "data_loader"/"transformer"/"data_exporter",
	"upstream_blocks": [
		"block_uuid1",
		"block_uuid2"
	],
	"config" : {
	        "data_source": "snowflake",
		"action_type": "remove",
		"axis": "column"
	}
    }
}
```
This `config` attribute holds all properties needed to generate templates. Currently there are 3 properties:
- `data_source` - specifies the source for a data loader. If the block is a data loader type, this property will be used to generate a template to load from this data source. Else, these properties are ignored.
- `action_type` and `axis` - specifies the transformer action to use in a transformer. If both of these properties are defined in the payload, they will be used to generate a template. If one or neither of these properties are defined, the default template is used.

Some other changes:
- Constants in the `models` package were refactored out into a separate constants file to avoid circular imports (since the 
templating system depends on constants from the `models` package but `block` and other models depend on the templating system)
- New constants were added to `transformer_actions` to track which actions require which parameters like `action_code`, `action_options`, and `outputs`

On a final note, some of the files in the `models` package were reformatted leading to large git blames, but actual code changes are much less.

# Tests
- Previous unit tests for the data preparation package were run in order to make sure that refactoring changes did not introduce errors
- New unit tests were developed to test template generation system (in the case of transformer actions)
- Local testing was performed to check whether block creation uses the correct template given the correct configuration.

cc:
@wangxiaoyou1993

### CODE REVIEW

File: mage_ai/data_cleaner/transformer_actions/constants.py [hunk 1]
```diff
@@ -70,7 +70,43 @@ class ActionType(str, Enum):
...
+        ActionType.MODE,
+        ActionType.SUM,
+        ActionType.AVERAGE,
+        ActionType.SHIFT_UP,
+        ActionType.SHIFT_DOWN,
+    ]
+)
+
+
+class Axis:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/transformer_actions/constants.py:109]: did you remove the `(str, Enum)`

File: mage_ai/data_preparation/templates/template.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    elif block_type == BlockType.TRANSFORMER:
+        template_source = __fetch_transformer_templates(config)
+    elif block_type == BlockType.DATA_EXPORTER:
+        template_source = __fetch_data_exporter_templates(config)
+    else:
+        template_source = ''
+    return template_source
+
+
+def __fetch_data_loader_templates(config: Mapping[str, str]):
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/templates/template.py:105]: nit: return str

File: mage_ai/data_preparation/templates/template.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        if additional_params_str != '':
+            additional_params_str = '\n        ' + additional_params_str
+        return (
+            template.render(action_type=action_type, axis=axis, kwargs=additional_params_str) + '\n'
+        )
+    else:
+        return template_env.get_template('transformers/default.py').render() + '\n'
+
+
+def __fetch_data_exporter_templates(config: Mapping[str, str]):
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/templates/template.py:139]: nit: return str

File: mage_ai/data_preparation/templates/template.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+import json
+import os
+import shutil
+
+template_env = jinja2.Environment(
+    loader=jinja2.FileSystemLoader('mage_ai/data_preparation/templates')
+)
+
+
+def build_template_from_suggestion(suggestion: Mapping) -> str:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/templates/template.py:19]: Do we call this when the suggested_action is in block config?

File: mage_ai/tests/data_preparation/test_templates.py [hunk 5]
```diff
@@ -1,4 +1,10 @@
-from mage_ai.data_preparation.templates.utils import build_template_from_suggestion
+from mage_ai.data_cleaner.transformer_actions.constants import ActionType, Axis
+from mage_ai.data_loader.base import DataSource
+from mage_ai.data_preparation.models.block import BlockType
+from mage_ai.data_preparation.templates.template import (
+    fetch_template_source,
```
Comment (wangxiaoyou1993) [inline mage_ai/tests/data_preparation/test_templates.py:5]: nit: alphabetical order

File: mage_ai/data_preparation/templates/template.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    action_type = config.get('action_type')
+    axis = config.get('axis')
+
+    if action_type is not None and axis is not None:
+        template = template_env.get_template('transformers/transformer_action_fmt.jinja')
+        additional_params = []
+        if action_type in ACTION_CODE_TYPES:
+            additional_params = ['action_code=\'your_action_code\'']
+        if action_type in ACTION_OPTION_TYPES:
+            additional_params.append('action_options={\'your_action_option\': None}')
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/templates/template.py:126]: You can add a TODO here to add the option keys given action_type.

Comment (skunichetty) [reply mage_ai/data_preparation/templates/template.py:133]: Added

File: mage_ai/tests/data_preparation/test_templates.py (hunk 5 above)
Comment (skunichetty) [reply mage_ai/tests/data_preparation/test_templates.py]: Fixed

File: mage_ai/data_preparation/templates/template.py (hunk 4 above)
Comment (skunichetty) [reply mage_ai/data_preparation/templates/template.py:19]: Changed so now if the suggested action payload is passed in `config`, then the custom template is built using that payload.

File: mage_ai/data_preparation/templates/template.py (hunk 3 above)
Comment (skunichetty) [reply mage_ai/data_preparation/templates/template.py]: Fixed.

File: mage_ai/data_preparation/templates/template.py (hunk 2 above)
Comment (skunichetty) [reply mage_ai/data_preparation/templates/template.py]: Fixed.

File: mage_ai/data_cleaner/transformer_actions/constants.py (hunk 1 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/transformer_actions/constants.py]: Fixed

File: mage_ai/data_preparation/models/block.py [hunk 7]
```diff
@@ -58,33 +47,37 @@ def downstream_block_uuids(self):
...
         uuid = clean_name(name)
         block_dir_path = os.path.join(repo_path, f'{block_type}s')
         if not os.path.exists(block_dir_path):
             os.mkdir(block_dir_path)
             with open(os.path.join(block_dir_path, '__init__.py'), 'w'):
                 pass
+
         file_path = os.path.join(block_dir_path, f'{uuid}.py')
         if os.path.exists(file_path):
-            if pipeline is not None and pipeline.has_block(uuid):
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/models/block.py:82]: Let's not delete this change.

Comment (skunichetty) [reply mage_ai/data_preparation/models/block.py:82]: Fixed

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**